File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/map/overlay/overlay/NMultipartPathOverlay.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay

import com.naver.maps.geometry.LatLng
import com.naver.maps.map.overlay.MultipartPathOverlay
import com.naver.maps.map.overlay.MultipartPathOverlay.ColorPart
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.AddableOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asBoolean
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asDouble
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asList
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.info.NOverlayInfo
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.NMultipartPath
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.NOverlayImage
import dev.note11.flutter_naver_map.flutter_naver_map.util.DisplayUtil

internal data class NMultipartPathOverlay(
    override val info: NOverlayInfo,
    val paths: List<NMultipartPath>,
    val widthDp: Double,
    val outlineWidthDp: Double,
    val patternImage: NOverlayImage?,
    val patternIntervalDp: Double,
    val progress: Double,
    val isHideCollidedCaptions: Boolean,
    val isHideCollidedMarkers: Boolean,
    val isHideCollidedSymbols: Boolean,
) : AddableOverlay<MultipartPathOverlay>() {


    override fun createMapOverlay(): MultipartPathOverlay = applyAtRawOverlay(MultipartPathOverlay())

    override fun applyAtRawOverlay(overlay: MultipartPathOverlay) = overlay.also { mo ->
        val coords = mutableListOf<List<LatLng>>()
        val colors = mutableListOf<ColorPart>()
        paths.forEach {
            coords.add(it.coords)
            colors.add(it.toColorPart())
        }
        mo.coordParts = coords
        mo.colorParts = colors
        mo.width = DisplayUtil.dpToPx(widthDp)
        mo.outlineWidth = DisplayUtil.dpToPx(outlineWidthDp)
        patternImage?.applyToOverlay(mo::setPatternImage)
        mo.patternInterval = DisplayUtil.dpToPx(patternIntervalDp)
        mo.progress = progress
        mo.isHideCollidedCaptions = isHideCollidedCaptions
        mo.isHideCollidedMarkers = isHideCollidedMarkers
        mo.isHideCollidedSymbols = isHideCollidedSymbols
    }

    companion object {
        fun fromMessageable(rawMap: Any): NMultipartPathOverlay = rawMap.asMap().let {
            NMultipartPathOverlay(
                info = NOverlayInfo.fromMessageable(it[infoName]!!),
                paths = it[pathsName]!!.asList(NMultipartPath::fromMessageable),
                widthDp = it[widthName]!!.asDouble(),
                outlineWidthDp = it[outlineWidthName]!!.asDouble(),
                patternImage = it[patternImageName]?.let(NOverlayImage::fromMessageable),
                patternIntervalDp = it[patternIntervalName]!!.asDouble(),
                progress = it[progressName]!!.asDouble(),
                isHideCollidedCaptions = it[isHideCollidedCaptionsName]!!.asBoolean(),
                isHideCollidedMarkers = it[isHideCollidedMarkersName]!!.asBoolean(),
                isHideCollidedSymbols = it[isHideCollidedSymbolsName]!!.asBoolean(),
            )
        }

        /*
            --- Messaging Name Define ---
        */
        private const val infoName = "info"
        const val pathsName = "paths"
        const val widthName = "width"
        const val outlineWidthName = "outlineWidth"
        const val patternImageName = "patternImage"
        const val patternIntervalName = "patternInterval"
        const val progressName = "progress"
        const val isHideCollidedCaptionsName = "isHideCollidedCaptions"
        const val isHideCollidedMarkersName = "isHideCollidedMarkers"
        const val isHideCollidedSymbolsName = "isHideCollidedSymbols"
        const val boundsName = "bounds"
    }
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/model/map/overlay/overlay/NMultipartPathOverlay.swift
import NMapsMap

internal struct NMultipartPathOverlay: AddableOverlay {
    typealias OverlayType = NMFMultipartPath
    var overlayPayload: Dictionary<String, Any?> = [:]

    let info: NOverlayInfo
    let paths: Array<NMultipartPath>
    let width: CGFloat
    let outlineWidth: CGFloat
    let patternImage: NOverlayImage?
    let patternInterval: Double
    let progress: CGFloat
    let isHideCollidedCaptions: Bool
    let isHideCollidedMarkers: Bool
    let isHideCollidedSymbols: Bool

    func createMapOverlay() -> OverlayType {
        let overlay = NMFMultipartPath()
        return applyAtRawOverlay(overlay)
    }
    
    func applyAtRawOverlay(_ overlay: NMFMultipartPath) -> NMFMultipartPath {
        paths.applyLineAndColor(linePartsFun: { overlay.lineParts = $0 }, colorPartsFun: { overlay.colorParts = $0 })
        overlay.width = width
        overlay.outlineWidth = outlineWidth
        overlay.patternIcon = patternImage?.overlayImage
        overlay.patternInterval = UInt(round(patternInterval))
        overlay.progress = progress
        overlay.isHideCollidedCaptions = isHideCollidedCaptions
        overlay.isHideCollidedMarkers = isHideCollidedMarkers
        overlay.isHideCollidedSymbols = isHideCollidedSymbols
        return overlay
    }

    static func fromMessageable(_ v: Any) -> NMultipartPathOverlay {
        let d = asDict(v)
        return NMultipartPathOverlay(
                info: NOverlayInfo.fromMessageable(d[infoName]!),
                paths: asArr(d[pathsName]!, elementCaster: NMultipartPath.fromMessageable),
                width: asCGFloat(d[widthName]!),
                outlineWidth: asCGFloat(d[outlineWidthName]!),
                patternImage: castOrNull(d[patternImageName], caster: NOverlayImage.fromMessageable),
                patternInterval: asDouble(d[patternIntervalName]!),
                progress: asCGFloat(d[progressName]!),
                isHideCollidedCaptions: asBool(d[isHideCollidedCaptionsName]!),
                isHideCollidedMarkers: asBool(d[isHideCollidedMarkersName]!),
                isHideCollidedSymbols: asBool(d[isHideCollidedSymbolsName]!)
        )
    }

    /*
        --- Messaging Name Define ---
    */
    private static let infoName = "info"
    static let pathsName = "paths"
    static let widthName = "width"
    static let outlineWidthName = "outlineWidth"
    static let patternImageName = "patternImage"
    static let patternIntervalName = "patternInterval"
    static let progressName = "progress"
    static let isHideCollidedCaptionsName = "isHideCollidedCaptions"
    static let isHideCollidedMarkersName = "isHideCollidedMarkers"
    static let isHideCollidedSymbolsName = "isHideCollidedSymbols"
    static let boundsName = "bounds"
}
